Client/Client.Core/Client.Core.Static/Client.cpp

```diff
@@ -657,4 +706,92 @@ void Client::onDirectMessageCreateAnswer(Utility::DirectMessageStatus directMess
         Base::Logger::LogLevel::WARNING
     );
 }
+
+void Client::onConnectionInfoAnswer(const Models::ConnectionInfo& connectionInfo)
+{
+    _connectionInfo = connectionInfo;
+
+    _connection->setConnectionVerifier(std::make_shared<Base::Verifiers::HashVerifier>());
+}
+
+void Client::onKeyAgreement(const Models::KeyAgreementInfo& serverKeyAgreementInfo)
+{
+    bool isSharedSecretCalculated = false;
+    CryptoPP::SecByteBlock sharedSecret;
+
+    if (!serverKeyAgreementInfo._publicKey.empty())
+    {
+        std::string publicServerKeyStr = serverKeyAgreementInfo._publicKey;
+        CryptoPP::SecByteBlock publicServerKey(reinterpret_cast<const CryptoPP::byte*>(publicServerKeyStr.data()), publicServerKeyStr.size());
+        sharedSecret = _connection->getKeyAgreement()->calculateSharedKey(publicServerKey);
+
+        isSharedSecretCalculated = !sharedSecret.empty();
+    }
+    else
+    {
+        Base::FileLogger::getInstance().log("Empty public server key. ", Base::LogLevel::WARNING);
+    }
+
+    if (isSharedSecretCalculated && !serverKeyAgreementInfo._publicKey.empty() &&
+        Base::SessionKeyHolder::Instance().setKey(std::move(sharedSecret), _connection->getUserID()) == Utility::GeneralCodes::SUCCESS)
```

Comment:
Kinda big if, maybe that's alright

Reply:
In case of failure of any check, the logic will be the same, so i think this is normal if.